forte/processors/data_augment/algorithms/skeleton_op.py

```diff
@@ -0,0 +1,765 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+SkeletonOp is the most low level Op used for Data Augmentation.
+Through this Op, users can use in-built utility functions to implement
+their own augmentation logic.
+"""
+from collections import defaultdict
+from copy import deepcopy
+from typing import (
+    DefaultDict,
+    Iterable,
+    List,
+    Tuple,
+    Union,
+    Dict,
+    Any,
+    Set,
+    cast,
+)
+from abc import abstractmethod, ABC
+from bisect import bisect_left, bisect_right
+from sortedcontainers.sorteddict import SortedDict
+from sortedcontainers.sortedlist import SortedList
+from forte.data.data_pack import DataPack
+from forte.common.configuration import Config
+from forte.data.span import Span
+from forte.data.ontology.top import (
+    Link,
+    Group,
+    MultiPackGroup,
+    MultiPackLink,
+    Annotation,
+)
+from forte.data.multi_pack import MultiPack
+from forte.data.ontology.core import BaseLink, Entry
+from forte.utils.utils import create_class_with_kwargs, get_class
+
+
+__all__ = ["SkeletonOp"]
+
+
+class SkeletonOp(ABC):
+    r"""
+    The SkeletonOp is the most basic augmentation Op that
+    gives users the most amount of freedom to implement their
+    logic of augmentation. The users are expected to use the
+    provided utility functions to implement their own augmentation
+    logic which will then be substantiated into new data packs. This
+    Op requires the users to have a relatively stronger understanding of
+    Forte's internal setup.
+    """
+
+    def __init__(self, configs: Union[Config, Dict[str, Any]]) -> None:
+
+        # :attr:`_replaced_annos`: {datapack id: SortedList[span, new text]}
+        # It records the spans replaced by new texts.
+        # It is a map from datapack id to a list of tuples
+        # (span, new text) inserted by :func:`replace`.
+        # The new text will be used for building new data pack.
+        self._replaced_annos: DefaultDict[
+            int, SortedList[Tuple[Span, str]]
+        ] = defaultdict(lambda: SortedList([], key=lambda x: x[0]))
+
+        # :attr:`_inserted_annos_pos_len`: {datapack id: Dict{position: length}}
+        # It records the position and length of inserted spans,
+        # mapping from datapack id to a dictionary (position -> length)
+        # inserted by :func:`insert`.
+        # The position is the index in the original datapack
+        # of insertion, and the length is the length of the inserted string.
+        self._inserted_annos_pos_len: DefaultDict[
+            int, SortedDict[int, int]
+        ] = defaultdict(lambda: SortedDict({}))
+
+        # :attr:`_deleted_annos_id`: {datapack id: Set[annotation tid]}
+        # It records the deleted span ids, mapping from datapack id
+        # to a set of annotation tids appended by :func:`delete`.
+        self._deleted_annos_id: DefaultDict[int, Set[int]] = defaultdict(set)
```

Comment:
Similar to above, here we use a slightly different data structure again. Maybe we can also store `SortedList[Tuple[int, int]]` so that we can quickly find overlaps?.

Reply:
Actually, this set up for recording deleted annotations is quite simple. Because when updating the new data pack, we just check that if the annotation which we are planning to add now to the data pack exists in the `deleted_annos_id` set or no. If it does, we just continue iterating, which essentially means that this annotation is not added.

Reply:
What `_replaced_annos` basically records is all types of operations done on a text, be it insert, delete or replace. the operation of deleting just a part of an annotation can be replicated as a collection of one insert and one delete task (we first delete the annotation and insert the part of it that we want to keep). This maintains a consistent structure. The job of creating the final data pack is determined by the `_replaced_annos` data structure.

Reply:
I think I understand a bit now, can you fix all the rest of the comments?